gpflow/kernels/periodic.py

```diff
@@ -65,20 +66,15 @@ def K_diag(self, X: tf.Tensor) -> tf.Tensor:
         return self.base_kernel.K_diag(X)
 
     def K(self, X: tf.Tensor, X2: Optional[tf.Tensor] = None) -> tf.Tensor:
-        if X2 is None:
-            X2 = X
-
-        # Introduce dummy dimension so we can use broadcasting
-        # TODO: does not support kernel broadcasting
-        f = tf.expand_dims(X, 1)  # now [N, 1, D]
-        f2 = tf.expand_dims(X2, 0)  # now [1, M, D]
-
-        r = np.pi * (f - f2) / self.period
+        r = np.pi * (difference_matrix(X, X2)) / self.period
         scaled_sine = tf.sin(r) / self.base_kernel.lengthscales
-        if hasattr(self.base_kernel, "K_r"):
-            sine_r = tf.reduce_sum(tf.abs(scaled_sine), -1)
-            K = self.base_kernel.K_r(sine_r)
+        if isinstance(self.base_kernel, IsotropicStationary):
+            if hasattr(self.base_kernel, "K_r"):
+                sine_r = tf.reduce_sum(tf.abs(scaled_sine), -1)
+                K = self.base_kernel.K_r(sine_r)
+            else:
+                sine_r2 = tf.reduce_sum(tf.square(scaled_sine), -1)
+                K = self.base_kernel.K_r2(sine_r2)
         else:
-            sine_r2 = tf.reduce_sum(tf.square(scaled_sine), -1)
-            K = self.base_kernel.K_r2(sine_r2)
+            K = self.base_kernel.K_d(scaled_sine)
```

Comment:
will this always form a positive-definite kernel ?

Reply:
I suppose not, it would be up to the user to make sure that this is a valid kernel. For now, it looks like it's not positive semidefinite with the cosine kernel. I can add some documentation for this

Reply:
The Periodic kernel is positive-definite for IsotropicStationary kernels, so perhaps it'd be better for now to simply say `Periodic` requires an IsotropicStationary (not generally Stationary) subclass as its base_kernel, and take out the special-casing for AnisotropicStationary ?